src/shaders/hillshade_prepare.fragment.glsl

```diff
@@ -0,0 +1,64 @@
+#ifdef GL_ES
+precision highp float;
+#else
+#define lowp
+#define mediump
+#define highp
+#endif
+
+uniform sampler2D u_image;
+varying vec2 v_pos;
+uniform vec2 u_dimension;
+uniform float u_zoom;
+
+float getElevation(vec2 coord, float bias) {
+    // Convert encoded elevation value to meters
+    vec4 data = texture2D(u_image, coord) * 255.0;
+    return (data.r + data.g * 256.0 + data.b * 256.0 * 256.0) / 4.0;
+}
+
+void main() {
+    vec2 epsilon = 1.0 / u_dimension;
+
+    // queried pixels:
+    // +-----------+
+    // |   |   |   |
+    // | a | b | c |
+    // |   |   |   |
+    // +-----------+
+    // |   |   |   |
+    // | d | e | f |
+    // |   |   |   |
+    // +-----------+
+    // |   |   |   |
+    // | g | h | i |
+    // |   |   |   |
+    // +-----------+
+
+    float a = getElevation(v_pos + vec2(-epsilon.x, -epsilon.y), 0.0);
+    float b = getElevation(v_pos + vec2(0, -epsilon.y), 0.0);
+    float c = getElevation(v_pos + vec2(epsilon.x, -epsilon.y), 0.0);
+    float d = getElevation(v_pos + vec2(-epsilon.x, 0), 0.0);
+    float e = getElevation(v_pos, 0.0);
+    float f = getElevation(v_pos + vec2(epsilon.x, 0), 0.0);
+    float g = getElevation(v_pos + vec2(-epsilon.x, epsilon.y), 0.0);
+    float h = getElevation(v_pos + vec2(0, epsilon.y), 0.0);
+    float i = getElevation(v_pos + vec2(epsilon.x, epsilon.y), 0.0);
+
+    vec2 deriv = vec2(
+        (c + f + f + i) - (a + d + d + g),
+        (g + h + h + i) - (a + b + b + c)
+        // decrease the value u_zoom is subtracted from to intensify terrain appearance
+    ) /  pow(2.0, 14.5 - u_zoom * 0.5);
```

Comment:
This number seemed a bit cryptic to me, so I separated out the calculations and made a suggestion for a tweaked value based on an arbitrary value for `a` that looked good to me:

Divide the differences by `8 * pixel size`
where pixel size (aka meters/pixel) is:  
`circumference of the world / (pixels per tile * number of tiles)`

`= 8 * 40075016.6855785 / (512 * pow(2, u_zoom))`
`= pow(2, 3) * pow(2, 25.25619978527) * pow(2, -9) * pow(2, -u_zoom)`
`= pow(2, 19.25619978527 - u_zoom)`

Multiply this number by a scale factor for vertical exaggeration by zoom level,
where the highest zl (z14) is the base and the exaggeration is proportional to zoom level: 
`pow(2, (u_zoom - 14) * a)`

`= pow(2, 19.25619978527 - u_zoom) * pow(2, u_zoom * a - (14 * a))`
`= pow(2, 19.25619978527 - (14 * a) + u_zoom * (a - 1))`

when `a = 0.3`:
`= pow(2.0, 15.06 - u_zoom * 0.7)`

Reply:
brilliant – thank you for the awesome breakdown @nickidlugash. I implemented what you suggest with a slight variation because there was barely any hillshading showing up at low zoom levels. take a look here https://github.com/mapbox/mapbox-gl-js/pull/5286/files#diff-2869d624151d9150a5f8f68cf567cb6fR61